Post: What's new on Apple Machine Learning at WWDC 2020? (#132)
xmartlabs/blog: _posts/2020-07-10-what s-new-on-apple-machine-learning-at-wwdc-2020.md

---
layout: post
title: What's new on Apple Machine Learning at WWDC 2020?
date: 2020-07-13 10:00:00

> author: Nicolas Lantean

Reviewer: ```suggestion
author: Nicolás Lantean
```

excerpt: "Discover the new updates on Apple Machine Learning frameworks at WWDC 2020"

> tags: [Mobile, Machine learning, iOS, Apple, WWDC2020]

Reviewer 2: Machine learning, CoreML, CreateML, Vision, WWDC20?

author_id: nicolantean
featured_image: /images/whats-new-on-apple-machine-learning-at-wwdc-2020/0-preview.png
show: true
category: development
---

# Introduction

Every day we have greater incorporation of machine learning tools into the mobile world. Apple knows this and every year it improves its frameworks in different ways and it also allows the ML integration into their platforms to become easier and easier, to the point that we don't even need to know machine learning fundamentals to start using these capabilities. This year was no exception since Apple announced some interesting new features on its machine learning frameworks and that's what I am gonna talk about in this blog.

> 

Reviewer 2: I am gonna `talk` about in this blog. -> I am gonna `cover` about in this blog.

> Every day we have greater incorporation of machine learning tools into the mobile world. Apple knows this and every year it improves its frameworks in different ways and it also allows the ML integration into their platforms to become easier and easier, to the point that we don't even need to know machine learning fundamentals to start using these capabilities. This year was no exception since Apple announced some interesting new features on its machine learning frameworks and that's what I am gonna cover in this blog.

Reviewer: I removed the "into the mobile world" stuff since I understand that's apple's improvement of integration in frameworks is what bridges them to the mobile world. I also rewrote  a fair number of sentences to improve clarity

```suggestion
Every day we more and more machine learning tools at our disposal. Apple knows this and every year it improves its frameworks in different ways, also allowing integrating ML on their platforms to become easier and easier, up to the point that we don't even need to know machine learning fundamentals in order to start using these capabilities. This year was no exception since Apple announced some interesting new features on its machine learning frameworks and that's what we are going to cover in this blog.
```

At Xmartlabs we have been closely involved in the field of on-device Machine Learning. Our main focus has been Machine Learning on iOS and we have been exploring different use cases including style-transfer, human body pose detection and medical imaging classification. Not only have we built some cool almost magical apps, but we have also rolled our sleeves  to create first of its kind development tools like **[Bender](https://xmartlabs.github.io/Bender/)** a few days before WWDC 17.

> 

Reviewer 2: ```suggestion

So now let's get into the new capabilities released by Apple at the WWDC 2020.
```

# Vision Framework

> The Vision framework provides solutions to computer vision challenges through a consistent interface, it performs face and face landmark detection, text detection, barcode recognition, image registration, and general feature tracking. Vision also allows the use of custom Core ML models.

Reviewer 2: I would recommend to move this to inside, what's new? section. I don't like `Vision Framework` title at the very beginning

> ## What's new?

Reviewer 2: ```suggestion
## What's new in Vision Framework?
```

Vision also allows the use of custom Core ML models.

Apple added two new capabilities to the Vision framework, both to detect human poses, one that recognizes body poses and one that recognizes the hand and its gestures.

The way Vision works is that you create a **VNRequest** object, which describes the task you want to perform and then you use a **VNImageRequestHandler** to execute it, so for these new features, they added two new requests.

### **VNDetectHumanBodyPoseRequest**

To perform a body pose request provides its **body pose-detection** capabilities through **VNDetectHumanBodyPoseRequest.**

The result of human body detection, is a struct data with 19 unique body points as shown in the figure below.

<div style="text-align: center"><img width="60%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/1-body-pose.png" /><p class="small">A human body with all the 19 unique body points illustrated.</p></div>

There are different applications that might be interested in using this data.

> For example, in a **[Swing Profile Golf Analyzer](https://apps.apple.com/us/app/swing-profile-golf-analyzer/id1039981052)** app, you can obtain the form of the person at the exact time of the shot, and then analyze it to show how well it was and make suggestions to improve the technique.

Reviewer: ```suggestion
For example, in a **[Swing Profile Golf Analyzer](https://apps.apple.com/us/app/swing-profile-golf-analyzer/id1039981052)** app, you can obtain the form of the person at the exact time of the shot, and then do analysis to rank how well it was, and what suggestions could improve the technique.
```

> Another VNDetectHumanBodyPoseRequest application could be to use the landmark points of a person's body to animate a caricature with its movements giving the possibility to create entertaining content.

Reviewer: In VNDetectHumanBodyPoseRequest bold or code is missing, right?

> The following snippet shows an example of how easy it is to perform the request for an image so you have a better sense of how easy it is to incorporate these pose detection ML functionality into your app:

Reviewer: Either go with functionalities in plural or change "these" for "this"
```suggestion
The following snippet shows an example of how easy it is to perform the request for an image. So you can have a better sense of how easy it is to incorporate these pose detection ML functionalities into your app:
```

        .bodyLandmarkKeyRightHip,
        .bodyLandmarkKeyRoot,
        .bodyLandmarkKeyLeftHip,
        .bodyLandmarkKeyLeftShoulder
    ]
    
    // Retrieve the CGPoints containing the normalized X and Y coordinates.
    let imagePoints: [CGPoint] = torsoKeys.compactMap {
        guard let point = recognizedPoints[$0], point.confidence > 0 else { return nil }
        
        // Translate the point from normalized-coordinates to image coordinates.
        return VNImagePointForNormalizedPoint(point.location,
                                              Int(imageSize.width),
                                              Int(imageSize.height))
    }
    
    // Draw the points onscreen.
    draw(points: imagePoints)
}
```

### **VNDetectHumanHandPoseRequest**

Similar to the body pose request, you can perform a **hand pose-detection** through **VNDetectHumanHandPoseRequest.**

The response provides 21 landmarks in total, four landmarks for each finger and one for the wrist.

<div style="text-align: center"><img width="60%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/2-hand-pose.png" /><p class="small">A hand with the 21 landmarks illustrated.</p></div>

> If you like to detect more than one hand, you have to set the **maximumHandCount** to the number of hands you want to detect.

Reviewer: ```suggestion
If you'd like to detect more than one hand, you can set the **maximumHandCount** to the number of hands you wish to detect.
```

Consider that this variable will affect the performance of your app, especially on older devices.

The code to perform a request and process the result data is analog to the body pose detection unlike the name of the request and the body landmarks keys.

## Use Model Deployment and Security with Core ML

### Model Deployment

> Apple also announced Model Deployment which allows developers to update the CoreML model without the need to update the app.

Reviewer: ```suggestion
Apple also announced Model Deployment which allows developers to update their CoreML model without the need of updating the app.
```

Before Model Deployment, developers were forced to submit an app update even though the only update was the machine learning model.

> As you may know, every submit to App Store requires an Appleeview which takes time.

Reviewer: ```suggestion
As you may know, every submit to the App Store requires an Apple review which takes time.
```

> Using Model Deployment machine learning specialists are able to keep improving the model accuracy and don't need to care about when the next app version comes out.

Reviewer: ```suggestion
Using Model Deployment, machine learning specialists are able to keep improving the model accuracy and not needing to care about when the next app version comes out.
```

> Model Deployment allows you to develop your model, adding new classes or retraining with more data for instance, without thinking in the updates plans of your app.

Reviewer: ```suggestion
Model Deployment allows you to develop your model, adding new classes or retraining with more data for instance. All this without taking into account any updating plan of your app.
```

> You can have two different calendars for updates which will be independent of each other.

Reviewer: ```suggestion
You can have two different calendars for updates, which will be independent of each other.
```

> As the model is uploaded in the cloud, the first time you try to get it with the new Core ML API, the model will try to download it in background.

Reviewer: ```suggestion
As the model is uploaded in the cloud, the first time you try to get it with the new Core ML API, the app will try to download the new model in background.
```

<div style="text-align: center"><img width="70%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/3-without-model-deployment.png" /><p class="small">Without Model Deployment.</p></div>

<div style="text-align: center"><img width="70%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/4-with-model-deployment.png" /><p class="small">With Model Deployment.</p></div>

To prepare your model for Model Deployment Xcode has now an option to create a Model Archive from your CoreML model. Then you only have to upload the archive through the Model Deployment Dashboard.
Additionally, before making the deployment you can add some targeting rules to assign different models to specific device class (iPhone / iPad / TV / Watch).
For example, the iPad's screen is bigger than the iPhone's one, so if you are working with a model that gets a drawing as an input, you can create two models in order to train them with the correct inputs.

That's good because your app doesn't have to store all the models for the different devices, reducing the app download size from the AppStore. 

> But what about the security of my models when being updated?

Reviewer: ```suggestion
But what about the security of my models when they are updated?
```

### Security with Core ML

> In regards to CoreML models security, Xcode is now able to encrypt your model at build time.

Reviewer: ![image](https://user-images.githubusercontent.com/42361379/89209240-46ec7880-d594-11ea-8c4c-742df39313c3.png)

> With this new feature, you can distribute your models knowing that they are end to end encrypted.

Reviewer: ```suggestion
With this new feature, you can distribute your models knowing that they are encrypted end to end.
```

But how do we take this new encryption capabilities to our app?

#### **How to encrypt and secure CoreML models in Xcode 12**

First, create a key (.mlmodelkey file) for the model in the **Utilities** section at the model file in Xcode. 

<div style="text-align: center"><img width="85%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/6-model-key.png" /></div>

Then, to encrypt your model you can add `—encrypt <path to .mlmodelkey>` to the model's Compiler Flags, or if you are using the new Model Deployment service, you need to use this key while creating the Model Archive.

In order to get the model key from the Apple servers to decrypt it later, the first time the app loads the model requires an internet connection, so it is highly recommended to use the new async load method shown in the code snippet below, which allows you to handle model errors.

```swift
FlowerStylizer.load { [self] result in
    switch result {
    case .success(let model):
        flowerStylizer = model
        DispatchQueue.main.async {
            applySlyledEffect(using: model)
        }
    case failure(let error):
        hanldeFailure(for: error)   
    }
}
```

> So we learned a bunch from security so far but what about building our own CoreML models? That's where CreateML (Apple framework designed to easily build machine learning models using Swift and Xcode) comes in...

Reviewer: ```suggestion
We learned a bunch from security so far, but what about building our own CoreML models? That's where CreateML (Apple framework designed to easily build machine learning models using Swift and Xcode) comes in...
```

## Build new CoreML models using **Create ML**

### **Build an Action Classifier**

There are some new models you can train in Create ML, for example, if you want to classify different types of actions from videos now you can do it.
This new feature is powered by Vision's body pose estimation, so humans are the target of these actions, not animals or objects. 

<div style="text-align: center"><img width="70%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/7-action-classifier.png" /><p class="small">How to create an Action Classifier.</p></div>

> The videos have to have one action type per video, several seconds per video and, only one person, framing the entire body.

Reviewer: ```suggestion
The videos must have one action type per video, several seconds per video and, only one person, and framing the entire body.
```

It's a good practice to train the model with an extra class called **Other**, where you place the videos in which the person is doing nothing and you don't want to recognize the action to prevent obtaining false results (i.e: people walking or just standing).

Before we start training the model, there are few parameters you can tweak.
Action duration represents the length of the action you try to recognize.
You have to analyze your actions to recognize and set that parameter correctly to obtain better results on your model.
Also, there is a horizontal flip option for data augmentation to increase the training data without adding new videos which use the flipped version as a train data apart of the original one.

Since the model takes a prediction window with poses of the frames recorded in the last X seconds set as action duration, as input, you have to make **VNDetectHumanBodyPoseRequest**, to extract these poses.

<div style="text-align: center"><img width="80%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/8-action-classifier-making-prediction.png" /><p class="small">Steps for making a prediction.</p></div>

### Build Image and Video Style Transfer models

> Another new type of model is the Image and Video Style Transfer model (a model based on two input images, one representing the artistic style and one representing the content).

Reviewer: ```suggestion
Another new type of model is the Image and Video Style Transfer model (a model based on two input images, one that consists of the artistic style, and another that's the content).
```

> After selecting that option on Create ML, you only have to set a Training Style Image which gives the style patterns you want to transfer, a Validation Image to validate the transfer while the trining is running and Content Images to train the model with images that represent the type of images you will use in your app.

Reviewer: What's "trining"? Training? Trimming?

```suggestion
After selecting the option on Create ML, you only have to set a Training Style Image which provides the style patterns you want to transfer, a Validation Image to validate the transfer while the training is running, and Content Images to train the model with images that represent the type of images you will use in your app.
```

> The default settings works very well but you can improve them.

Reviewer: ```suggestion
The default settings work pretty well, but you can improve them.
```

In order to do that you have to select if you will use the model for images or videos, then you can select the number of iterations in the training which has two parameters that will define your model.
Style Strength indicates how much style the model will transfer from the image and Style Density determines how much detail of the style the model will transfer.

<div style="text-align: center"><img width="90%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/9-style-transfer-model.png" /><p class="small">Style and Stylized result with a high Style Strength and low Style Density.</p></div>

### Control training in Create ML with Swift

> Although Create ML gives us a very straightforward way to create Core ML models, it's very limited for more sophisticated machine learning models.

Reviewer: ```suggestion
Although Create ML provides us with a very straightforward way to create Core ML models, it's very limited for more sophisticated machine learning models.
```

> In order to overcome that, Apple adds the possibility to control training through Swift code to save time and obtain better results while creating a new model.

Reviewer: ```suggestion
In order to overcome that, Apple adds the possibility to control training through Swift code, in order to save time and obtain better results while creating a new model.
```

> The best thing to do is to use Xcode Playground where it instantly shows the results of the code that you write.

Reviewer: We could add [links](url) to all these things such as Xcode Playground, and all the apple services, right?

With the new API, you can call a training method that will train the model and return a job.
Before doing that, you have to specify the session parameters.
These parameters consist of **sessionDirectory**, **reportInterval**, **checkpointInterval**, and **iterations**.
The job returned in the training method, contains progress, checkpoint, and result publishers.
It also has a cancel method that allows you to stop the training at any point.
Using the result and progress publishers, you can handle success and errors, obtain the resulting model, and also observe how the training process is going.

<div style="text-align: center"><img width="65%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/10-control-training.png" /><p class="small">Checkpoint availability on different model types.</p></div>

Now with checkpoints, you can capture the state of your model over time.

> One benefit of that is when your model stop training but you notice that the accuracy is still growing up, so you can resume the training increasing the number of iterations without having to train the model again from scratch. 

Reviewer: ```suggestion
One benefit of that is when your model stop training but you notice that the accuracy is still growing up, so you can resume the training increasing the number of iterations without having to re-train the model again from scratch. 
```

### Get models on device using Core ML Converters

Create ML is an intuitive and easy tool for training models but doesn't cut it for complex projects with models that are not available on the framework.
In these cases, often we use another tool, like TensorFlow or PyTorch, but these tools can't create .mlmodel by themselves to use in Core ML.

> That's where coremltools (we have taken about it in the blog [How to convert a NN model from TensorFlow Lite to CoreML](https://blog.xmartlabs.com/2019/11/22/TFlite-to-CoreML/)) appears, an Apple framework that allows you to export your model in .mlmodel format.

Reviewer: ```suggestion
That's where coremltools (which we previously discussed it in the blog [How to convert a NN model from TensorFlow Lite to CoreML](https://blog.xmartlabs.com/2019/11/22/TFlite-to-CoreML/)) appears, an Apple framework that allows you to export your model in .mlmodel format.
```

This year Apple focused on the two most commonly used frameworks, TensorFlow and PyTorch. To convert TensorFlow, tf.keras, and PyTorch models, now you have to use the new single conversion API with its new unified interface.

<div style="text-align: center"><img width="65%" src="/images/whats-new-on-apple-machine-learning-at-wwdc-2020/11-coremltools4.png" /><p class="small">New unified convert() method.</p></div>

What the new **convert** function is doing under the hood is to inspect the model format and choose the correct converter for it.
Then it converts the model into an intermediate representation called **MIL**.
Having the model in a **MIL** format gives us the possibility to optimize the model in different ways, for example removing unnecessary operations.
The interesting part of **MIL** is that allows you to deal with layers that are not directly supported by Core ML, you can split it up into more primitive MIL operations, like matrix multiplications, or other arithmetic.

## My final thoughts

Apple has done a really great job improving its machine learning frameworks. 

> **Vision** with the new body and hand pose detectors brings a huge of possibilities to your app like recognize different types of gestures and do some cool stuff like take pictures without pressing a button or analyze body poses doing different sports to improve your form on some technique. 

Reviewer: ```suggestion
**Vision** with the new body and hand pose detectors brings a huge of possibilities to your app like recognizing different types of gestures and do some cool stuff like taking pictures without pressing a button or analyzing body poses when doing sports in order to improve your form on some technique. 
```

By adding Model Deployment and model encryption, **CreateML** will make developers feel more comfortable using it.
It will help them save time by avoiding to update their app frequently only to modify its models and it will make them feel more secure to use models not available on the machine learning community which has a great value without the risk of being stolen.

**Create ML** keeps extending the available models with the addition of Action classifier and style transfer models.

> And talking about control training and Core ML converters, Create ML is less limited than before, and **coremltools** with its new version allows developers to convert its customs models created in other tools easier.

Reviewer: Isn't this ending kind of dry?

I'd go with a paragraph that sums up our opinion (a step in the right direction) and maybe tease to a future blog we have on the pipeline that relates to this.

Apple has done a really great job improving its machine learning frameworks. 

**Vision** with the new body and hand pose detectors brings a huge of possibilities to your app like recognizing different types of gestures and do some cool stuff like taking pictures without pressing a button or analyzing body poses when doing sports in order to improve your form on some technique.

By adding Model Deployment and model encryption, **CreateML** will make developers feel more comfortable using it.
It will help them save time by avoiding to update their app frequently only to modify its models and it will make them feel more secure to use models not available on the machine learning community which has a great value without the risk of being stolen.

**Create ML** keeps extending the available models with the addition of Action classifier and style transfer models.

And talking about control training and Core ML converters, Create ML is less limited than before, and **coremltools** with its new version allows developers to convert its customs models created in other tools easier.

To sum this up, here are some thoughts on the great new features and improvements on Apple's frameworks:

> * Machine learning apps no longer need deep expertise in the feed, nor to train some basic models.

Reviewer: ```suggestion
* Machine learning apps no longer need deep expertise in the field, some basic models can even be trained without it.
```
